src/common/promiseUtils.ts

```diff
@@ -182,3 +182,65 @@ export async function formatAggregateError<T>(
     throw new AggregateError(errors, message)
   })
 }
+
+export async function asyncStaggeredRace(
```

Comment:
How does this differ from `asyncWaterfall`? I'm pretty sure it's trying to do the exact same thing, although I do like this implementation a lot better.

Reply:
The difference between the two is asyncWaterfall throws away good work that isn't done before the timeout